### PR TITLE
Tooling updates for string de-duplication

### PR DESCRIPTION
**Pull request by ydnar**
This PR implements code to build a global string table in the code generation part of the `internal/build` package.

Due to small changes in zone metadata resulting in large diffs in the generated `zones.go` file, the usage of the string table is currently disabled. 

We can revive it, fully or partially in the future. I’d like to merge this PR so this code doesn’t rot in a branch.

### CODE REVIEW

File: internal/build/generate.go [hunk 1]
```diff
@@ -46,22 +47,41 @@ func GenerateGo(zones map[string]*Zone) error {
...
+		data.domainStringSlice(zones[d].NameServers)
+	}
+	for _, d := range domains {
+		data.domainStringSlice(zones[d].Wildcards)
+	}
+	for _, d := range domains {
+		data.domainString(zones[d].WhoisServer)
+	}
+	for _, d := range domains {
+		data.urlString(zones[d].WhoisURL)
```
Comment (case) [inline internal/build/generate.go:75]: We'll need to get the forthcoming RDAP server URLs into this list as well.

Comment (ydnar) [reply internal/build/generate.go:75]: That’s the idea.

**Comment by chriskillpack** [comment]
What amount of memory savings does this interning produce?

File: internal/build/strings.go [hunk 2]
```diff
@@ -97,3 +97,32 @@ func NormalizeDomains(in []string) []string {
...
 }
+
+// IndexOfStrings finds a rank-sorted slice of domain names (needle)
+// within a larger slice (haystack).
+func IndexOfStrings(haystack []string, needle []string) int {
+outer:
+	for i := range haystack {
+		for j := range needle {
+			if i+j >= len(haystack) || needle[j] != haystack[i+j] {
+				continue outer
```
Comment (chriskillpack) [inline internal/build/strings.go:108]: TIL https://golang.org/ref/spec#Continue_statements

File: internal/build/generate.go [hunk 3]
```diff
@@ -200,20 +270,26 @@ var y = [{{len .Zones}}]Zone{
...
 			{{if $z.SubdomainsEnd}} z[{{$z.SubdomainsOffset}}:{{$z.SubdomainsEnd}}] {{else}} x {{end}}, \
-			{{if $z.NameServers}} s{ {{range $z.NameServers}}{{quotedDomain .}},{{end}}} {{else}} n {{end}}, \
-			{{if $z.Wildcards}} s{ {{range $z.Wildcards}}{{quoted .}},{{end}}} {{else}} n {{end}}, \
-			{{if $z.Locations}} s{ {{range $z.Locations}}{{quoted .}},{{end}}} {{else}} n {{end}}, \
-			{{if $z.Languages}} s{ {{range $z.Languages}}{{quoted .}},{{end}}} {{else}} n {{end}}, \
-			{{quotedDomain $z.WhoisServer}}, \
-			{{quotedURL $z.WhoisURL}}, \
-			{{quotedURL $z.InfoURL}}, \
+			{{domainSlice $z.NameServers}}, \
+			{{domainSlice $z.Wildcards}}, \
```
Comment (chriskillpack) [inline internal/build/generate.go:278]: In the removed code `Wildcards` was not treated as a domain, it used `quoted` not `quotedDomain`, but in this change `Wildcards` uses `domainSlice` and not `stringSlice`. Is that inentional?

File: internal/build/generate.go (hunk 1 above)
Comment (chriskillpack) [reply internal/build/generate.go:75]: Yup, will handle that separately once the RDAP support is merged. It's a trivial add.

File: internal/build/generate.go (hunk 3 above)
Comment (ydnar) [reply internal/build/generate.go:298]: Wildcards are either IP addresses or hostnames, so should be escaped/normalized as domain names.

**Comment by ydnar** [comment]
One issue with this PR is that small changes to metadata will create big diffs in `zones.go`.

**Pull request merged**